---
title: "RNA Seq Variant Calling"
author: "Amy Paguirigan"
date: "12/9/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Install Required Packages

## For Local Use
```{r eval=FALSE, message=FALSE, warning=TRUE, include=FALSE}
remotes::install_github('FredHutch/tgR')
remotes::install_github('FredHutch/fh.wdlR')
```

# Load Packages
```{r}
library(tidyverse); library(aws.s3); library(tgR); library(fh.wdlR)
```

# Set Credentials (run one line)
```{r eval=FALSE, include=FALSE}
setCreds(tokenSet = "file", path = "~/github/cred/paguirigan.R") #DVTC
```

# Find files
```{r}
tags <- listS3Objects(bucket = "fh-pi-paguirigan-a")
annotations <- tgrAnnotate(DAG = "paguirigana")
monsterMash <- dplyr::left_join(tags, annotations) %>% dropWhen()

rna <- monsterMash %>% filter(genomics_type == "rnaseq" & is.na(workflowID) == T & 
                                  specimen_privacy == "publiclyavailable"); rna %>% group_by(stage) %>% summarize(n_distinct(molecular_id))

k562 <- rna %>% filter(omics_sample_name == "K562")
```

```{r}
#Manually made a batch file this first time. 
```


```{r}

```


# Submit Job to gizmo Cromwell
Note: setwd to location of workflow files
```{r}
valid <- womtoolValidate(WDL = "SC10x-gizmo-umisplit.wdl"); valid[["errors"]]
thisJob <- cromwellSubmitBatch(WDL = "SC10x-gizmo-umisplit.wdl",
                    Params = "SC10x-gizmo-parameters.json",
                    Batch = "SC10x-gizmo-sample.json",
                    Options = "workflow-options.json",
                    Labels = data.frame("workflowNotes" = "umisplit-lessSTARmemory"))

thisOne<- thisJob$id; thisOne
```

```{r}
```


# Monitor Running Jobs
```{r}
jobs <- cromwellJobs()
WTF <- cromwellGlob(thisOne)
w <- cromwellWorkflow(thisOne); w$status
c <- cromwellCall(thisOne); c %>% group_by(callName, executionStatus) %>% summarize(status = n())%>% arrange(executionStatus)
ca <- cromwellCache(thisOne); ca %>% group_by(callCaching.hit, callName) %>% summarize(hits = n()) 
cromwellTiming(thisOne)
butWhy <- left_join(c, mutate_all(ca, as.character)); butWhy %>% group_by(callName, executionStatus, callCaching.hit) %>% summarize(hits = n()) %>% arrange(desc(executionStatus), callCaching.hit)
f <- cromwellFailures(thisOne)
#abort <- cromwellAbort(thisOne) # Careful with this
```


# Output Processing workflow prep (copyNTag)
```{r}
out <- cromwellOutputs(thisOne)
#batchFileName <- gsub("tsv.*$", "tsv", gsub("^.*batchFile[^s3]*s3://fh-pi-paguirigan-a-genomicsrepo/", "", w$inputs))
batchFile <- s3read_using(FUN = read.delim, stringsAsFactors = F,
                           object = batchFileName,
                           bucket = "fh-pi-paguirigan-a-genomicsrepo")
#batchFile <- manifest
batchFile$shardIndex <- as.character(seq(from = 0, to = nrow(batchFile)-1, by = 1))
annotatedOutputs <- inner_join(batchFile, out, by = c("shardIndex")) 
```




